File: localizator.py
from __future__ import print_function

import argparse
import os
import os.path
import pickle

import translations
import urllib3
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

try:
    import apiclient
except ImportError:
    print("'import apiclient' failed; run: pip3 install google-api-python-client\n"
          "or manually on https://developers.google.com/api-client-library/python/start/installation")
    exit(-1)

try:
    import oauth2client
    from oauth2client import client
    from oauth2client import tools
    from oauth2client import file
except ImportError:
    print("'import oauth2client' failed; run: pip3 install oauth2client")
    exit(-1)

try:
    parser = argparse.ArgumentParser(description='Create localizable files')
    parser.add_argument('--id', help='provide file id to avoid prompt')
    parser.add_argument('--path', help='Path destination for *.lproj folders', default='./')
    parser.add_argument('--platform', choices=['ios', 'android'], help='Should be either ios or android', default='ios')
    parser.add_argument('--gid', help='Use the Google sheet ID from the end of the url link')
    parser.add_argument('--keep_csv', type=bool, help='Should keep the CSV file on the disk', default=False)
    args = parser.parse_args()
    flags = args
except ImportError:
    flags = None
    print("Cannot parse")

SCOPES = ['https://www.googleapis.com/auth/drive']
CLIENT_SECRET_FILE = 'client_secret.json'
APPLICATION_NAME = 'YOUR_PROJECT_NAME_GOES_HERE'


def get_credentials():
    creds = None
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                CLIENT_SECRET_FILE, SCOPES)
            creds = flow.run_local_server(port=0)
        # Save the credentials for the next run
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)
    return creds


def get_files(service):
    results = service.files().list(fields='files(id, exportLinks)').execute()
    items = results.get('files', [])
    if not items:
        print('No files found.')
    else:
        return service, items


def main():
    if not os.path.exists(CLIENT_SECRET_FILE):
        print("'" + CLIENT_SECRET_FILE + "' file does not exist.\n"
                                         "Enable the API and save the '" + CLIENT_SECRET_FILE + "' file here.\n"
                                                                                                "https://console.developers.google.com/flows/enableapi?apiid=drive")
        exit(-1)

    credentials = get_credentials()
    print("Authorizing...")
    service = build('drive', 'v3', credentials=credentials)
    print("Listing files...")
    service, files = get_files(service)

    if args.id:
        for item in files:
            if item['id'] == args.id:
                file = item
        if not 'file' in locals():
            print("File with id '" + args.id + "' not found in files list.\n"
                                               "Notice: you need to load the file at least once in your browser, so it's visible in your files list.")
            exit(-1)
    else:
        i = 0
        for item in files:
            print("[" + str(i) + "] " + str(item['title']) + " - " + str(item['id']))
            i += 1
        isDigit = False
        while not isDigit:
            index = input("Select a file index: \n")
            isDigit = index.isdigit()
            if int(index) > len(files) or int(index) < 0:
                print("Invalid index supplied. Try again")
                isDigit = False
        file = files[int(index)]

    content = download_file(service, file, args.gid)
    filename = 'tmp.csv'
    csvf = open(filename, 'w')
    csvf.write(content.decode("utf-8"))
    csvf.close()

    if args.platform == 'ios':
        translations.translate(filename, args.path)
    elif args.platform == 'android':
        translations.translate_android(filename, args.path)
    else:
        print("Invalid platform. type --help for help")
    if not args.keep_csv:
        os.remove(filename)
    print("Success! Localization files for platform " + args.platform + " have been generated under '" + args.path + "'")


def download_file(service, drive_file, tab_id):
    download_url = drive_file['exportLinks']['text/csv'] + "&gid=" + tab_id
    print("Downloading " + download_url)
    resp, content = service._http.request(download_url)
    if resp.status == 200:
        return content
    else:
        print('HTTP error: %s' % resp.status)
        if args.gid:
            print('Maybe --gid is wrong?')
        exit(-1)


if __name__ == '__main__':
    main()

File: localizator.js
// Co-created by Livio Gamassia and Celian Moutafis: https://github.com/LivioGama/localizator
// https://developers.google.com/drive/api/quickstart/nodejs
import {authenticate} from '@google-cloud/local-auth'
import fs from 'fs'
import {google} from 'googleapis'
import path from 'path'
import process from 'process'
import yargs from 'yargs'

import {hideBin} from 'yargs/helpers'
import {translateWeb, translateAndroid, translateIos} from './translations.js'

const yargsInstance = yargs(hideBin(process.argv))
const args = yargsInstance.wrap(yargsInstance.terminalWidth()).parse()

const SCOPES = ['https://www.googleapis.com/auth/drive']
const TOKEN_PATH = path.join(process.cwd(), 'token.json')
const CREDENTIALS_PATH = path.join(process.cwd(), 'client_secret.json')

const getCredentials = async () => {
  let credentials = null
  try {
    const content = fs.readFileSync(TOKEN_PATH, 'utf-8')
    credentials = JSON.parse(content)
  } catch (err) {
    const client = await authenticate({
      scopes: SCOPES,
      keyfilePath: CREDENTIALS_PATH,
    })
    if (client.credentials) {
      const cs = JSON.parse(fs.readFileSync(CREDENTIALS_PATH, 'utf-8'))
      const payload = JSON.stringify({
        type: 'authorized_user',
        client_id: client.credentials.client_id || cs.installed.client_id,
        client_secret: client.credentials.client_secret || cs.installed.client_secret,
        refresh_token: client.credentials.refresh_token,
      })
      fs.writeFileSync(TOKEN_PATH, payload)
      credentials = JSON.parse(payload)
    }
  }
  return credentials
}

const getFiles = async service => {
  const res = await service.files.list({fields: 'files(id, name)'})
  const files = res.data.files
  if (files.length === 0) {
    console.log('No files found.')
    return null
  }
  return {service, files}
}

const downloadFile = async (service, driveFile, tabId) => {
  console.log(`Downloading ${driveFile.id}`)
  const res = await service.files.export({fileId: driveFile.id, mimeType: 'text/csv'})
  if (res.status === 200) {
    return res.data
  } else {
    console.log(`HTTP error: ${res.status}`)
    if (tabId) {
      console.log('Maybe --gid is wrong?')
    }
    process.exit(-1)
  }
}

const main = async () => {
  if (!fs.existsSync(CREDENTIALS_PATH)) {
    console.log(
      `'${CREDENTIALS_PATH}' file does not exist.\nEnable the API and save the '${CREDENTIALS_PATH}' file here.\nhttps://console.developers.google.com/flows/enableapi?apiid=drive`
    )
    process.exit(-1)
  }

  const credentials = await getCredentials()
  console.log('Authorizing...')
  const auth = google.auth.fromJSON(credentials)
  const service = google.drive({version: 'v3', auth})
  console.log('Listing files...')
  const {files} = await getFiles(service)

  let file
  if (args.id) {
    const fileId = args.id
    file = files.find(item => item.id === fileId)
    if (!file) {
      console.log(
        `File with id '${fileId}' not found in files list.\nNotice: you need to load the file at least once in your browser, so it's visible in your files list.`
      )
      process.exit(-1)
    }
  } else {
    files.forEach((item, index) => {
      console.log(`[${index}] ${item.name} - ${item.id}`)
    })
    const index = await new Promise(resolve => {
      process.stdin.resume()
      process.stdin.setEncoding('utf8')
      process.stdout.write('Select a file index: ')
      process.stdin.on('data', data => {
        const selectedIndex = parseInt(data.trim(), 10)
        if (selectedIndex >= 0 && selectedIndex < files.length) {
          resolve(selectedIndex)
        } else {
          console.log('Invalid index supplied. Try again')
          process.stdout.write('Select a file index: ')
        }
      })
    })
    file = files[index]
  }

  const tabId = args.gid || ''
  const content = await downloadFile(service, file, tabId)
  const filename = 'tmp.csv'
  fs.writeFileSync(filename, content)

  // retrieve from args a list of languages and transform them into an array
  const languages = (args.languages || 'en').split(',')
  console.log(`Languages: ${languages} (in Google Spreadsheet columns order)`)

  const platform = args.platform || 'web'
  const path = args.path || './'
  if (platform === 'ios') {
    translateIos(filename, path, languages)
  } else if (platform === 'android') {
    translateAndroid(filename, path, languages)
  } else if (platform === 'web') {
    translateWeb(filename, path, languages)
  } else {
    console.log('Invalid platform. type --help for help')
  }

  const keepCsv = args.keep_csv
  if (!keepCsv) {
    fs.unlinkSync(filename)
  }
  console.log(
    `Success! Localization files for platform ${platform} have been generated under '${path}'`
  )
}

main().catch(console.error)
